hooks/command

```diff
@@ -0,0 +1,236 @@
+  local prefix="$1"
+  local parameter="${prefix}_0"
+
+  if [[ -n "${!parameter:-}" ]]; then
+    local i=0
+    local parameter="${prefix}_${i}"
+    while [[ -n "${!parameter:-}" ]]; do
+      echo "${!parameter}"
+      i=$((i+1))
+      parameter="${prefix}_${i}"
+    done
+  elif [[ -n "${!prefix:-}" ]]; then
+    echo "${!prefix}"
+  fi
+}
+
+retry() {
+  local retries=$1
+  shift
+
+  local count=0
+  until "$@"; do
+    exit=$?
+    wait=$((retry_sleep ** count))
+    count=$((count + 1))
+    if [ $count -lt "$retries" ]; then
+      echo "Retry $count/$retries exited $exit, retrying in $wait seconds..."
+      sleep $wait
+    else
+      echo "Retry $count/$retries exited $exit, no more retries left."
+      return $exit
+    fi
+  done
+  return 0
+}
+
+function get_proc_type_image_tag() {
+  echo "registry.heroku.com/$app/$1:latest"
+}
+
+app=$(plugin_read_list APP)
+process_types=$(plugin_read_list PROCESS_TYPES)
+
+##
+## Pulling from owner container registry
+##
+
+for line in $process_types; do
+  IFS=':' read -r -a tokens <<< "$line"
+  proc_type=${tokens[0]}
+  proc_type_image=$(IFS=':'; echo "${tokens[*]:1}")
+  proc_type_image_id=$(docker images -q "$proc_type_image")
+
+  echo "~~~ :docker: Pulling $proc_type image"
+  if [ -z "${proc_type_image_id}" ]; then
+    status=0
+    if ! retry 3 docker pull "$proc_type_image"; then
+      status=$((status + 1))
+    fi
+    if [ $status -eq 0 ]; then
+      echo "Pulled $proc_type_image"
+    else
+      echo "Failed pull $proc_type_image"
+      exit 1
+    fi
+  else
+    echo "Found $proc_type_image"
+  fi
+done
+
+##
+## Tagging for heroku container registry
+##
+
+for line in $process_types; do
+  IFS=':' read -r -a tokens <<< "$line"
+  proc_type=${tokens[0]}
+  proc_type_image=$(IFS=':'; echo "${tokens[*]:1}")
+  proc_type_tag=$(get_proc_type_image_tag "$proc_type")
+
+  echo "~~~ :heroku: Tagging ${proc_type} image"
+  docker tag "$proc_type_image" "$proc_type_tag"
+  echo "Tagged $proc_type_image as $proc_type_tag"
+done
+
+##
+## Pushing to heroku container registry
+##
+
+echo "~~~ :heroku: Logging into Heroku Docker Registry"
+echo "$HEROKU_API_KEY" | docker login --username=_ --password-stdin registry.heroku.com
+
+for line in $process_types; do
+  IFS=':' read -r -a tokens <<< "$line"
+  proc_type=${tokens[0]}
+  proc_type_image=$(IFS=':'; echo "${tokens[*]:1}")
+  proc_type_tag=$(get_proc_type_image_tag "$proc_type")
+
+  echo "~~~ :heroku: Pushing ${proc_type} image"
+  status=0
+  if ! retry 3 docker push "$proc_type_tag"; then
+    status=$((status + 1))
+  fi
+  if [ $status -eq 0 ]; then
+    echo "Pushed $proc_type_tag"
+  else
+    echo "Failed push $proc_type_tag"
+    exit 1
+  fi
+done
+
+##
+## Releasing to heroku platform
+##
+
+proc_type_names=()
+for line in $process_types; do
+  IFS=':' read -r -a tokens <<< "$line"
+  proc_type=${tokens[0]}
+  proc_type_names+=("$proc_type")
+done
+
+echo "~~~ :heroku: Releasing ${proc_type_names[*]}"
```

Comment:
Looking at the docker-compose Buildkite plugin, the preparatory sections are logged with `~~~` and the main commands logged with `+++`. Consider logging with `+++` for some sections.